src/providers/ipa/ipa_deskprofile_rules_util.c

```diff
@@ -0,0 +1,917 @@
+    return ret;
+}
+
+errno_t
+ipa_deskprofile_rules_create_user_dir(
+                                    const char *username, /* fully-qualified */
+                                    uid_t uid,
+                                    gid_t gid)
+{
+    TALLOC_CTX *tmp_ctx;
+    char *shortname;
+    char *domain;
+    char *domain_dir;
+    errno_t ret;
+
+    tmp_ctx = talloc_new(NULL);
+    if (tmp_ctx == NULL) {
+        return ENOMEM;
+    }
+
+    ret = sss_parse_internal_fqname(tmp_ctx, username, &shortname, &domain);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "sss_parse_internal_fqname() failed [%d]: %s\n",
+              ret, sss_strerror(ret));
+        goto done;
+    }
+
+    ret = sss_create_dir(IPA_DESKPROFILE_RULES_USER_DIR, domain, 0755,
+                         getuid(), getgid());
+    if (ret != EOK) {
+        DEBUG(SSSDBG_TRACE_FUNC,
+              "Failed to create the directory \"%s/%s\" that would be used to "
+              "store the Desktop Profile rules users' directory [%d]: %s\n",
+              IPA_DESKPROFILE_RULES_USER_DIR, domain,
+              ret, sss_strerror(ret));
+        goto done;
+    }
+
+    domain_dir = talloc_asprintf(tmp_ctx, IPA_DESKPROFILE_RULES_USER_DIR"/%s",
+                                 domain);
+    if (domain_dir == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+
+    ret = sss_create_dir(domain_dir, shortname, 0600, uid, gid);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_TRACE_FUNC,
+               "Failed to create the directory \"%s/%s/%s\" that would be used "
+               "to store the Desktop Profile rules for the user \"%s\" [%d]: "
+               "%s\n",
+               IPA_DESKPROFILE_RULES_USER_DIR, domain, shortname, username,
+               ret, sss_strerror(ret));
+        goto done;
+    }
+
+    ret = EOK;
+
+done:
+    talloc_free(tmp_ctx);
+    return ret;
+}
+
+static errno_t
+ipa_deskprofile_get_normalized_rule_name(TALLOC_CTX *mem_ctx,
+                                         const char *name,
+                                         char **_rule_name)
+{
+    char *rule_name;
+    char buffer[1024];
+    size_t buffer_len;
+    size_t name_len;
+
+    name_len = strlen(name);
+    buffer_len = 0;
+    for (size_t i = 0; i < name_len; i++) {
+        char character;
+        bool replace;
+
+        character = name[i];
+        replace = false;
+
+        if (isalnum(character) == 0) {
+            char next_character;
+
+            next_character = name[i+1];
+            if (i + 1 >= name_len || isalnum(next_character) == 0) {
+                continue;
+            }
+
+            replace = true;
+        }
+
+        buffer[buffer_len] = replace ? '_' : character;
+        buffer_len++;
+    }
+    buffer[buffer_len] = '\0';
+
+    rule_name = talloc_strdup(NULL, buffer);
+    if (rule_name == NULL) {
+        return ENOMEM;
+    }
+
+    *_rule_name = talloc_strdup(mem_ctx, rule_name);
```

Comment:
You can just strdup to `_rule_name` directly and use the `mem_ctx`, dupinig to a NULL context and then stealing seems redundant.